Add a guide about upgrading applications without downtime

pleshakov opened the pull request:
### Proposed changes

Problem:
As a user of NKG, I want a guide to explain the different strategies I can upgrade my application (NOT NKG) while using NKG specific features (such as traffic splitting) So that I minimize the downtime my application encounters.

Solution:
- Add a guide that covers NKG specific features that help upgrade applications without downtime
- Introduce a new /docs/guide folder for guides.
- Add a link to the guides folder from the main README.

Testing: none

Closes https://github.com/nginxinc/nginx-kubernetes-gateway/issues/704

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/nginx-kubernetes-gateway/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto main
- [x] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

pleshakov commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+> By application downtime we mean the clients cannot get responses from your application. Instead,
+> they get responses with an error status code like 502 from NGINX.
+
+This guide covers:
```
What do you think about adding every case considered in this guide to our test automation, so that it checks that an upgrade happens without downtime?

sjberman replied:
Since we don't have test automation yet, do we have a list (or need a list) of potential tests?

pleshakov replied:
there is https://github.com/nginxinc/nginx-kubernetes-gateway/discussions/826 and also https://github.com/nginxinc/nginx-kubernetes-gateway/discussions/752

pleshakov replied:
added in https://github.com/nginxinc/nginx-kubernetes-gateway/discussions/826#discussioncomment-6705198

sjberman commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+This guide shows how the specific features of NGINX Kubernetes Gateway can help upgrade your applications without
+downtime.
+
+> We assume you are familiar with the upgrade methods mentioned in this guide. As a consequence, we will not cover
+> them in depth, but instead, focus on how NGINX Kubernetes Gateway supports them.
```
```suggestion
> We assume you are familiar with the upgrade methods mentioned in this guide. As a result, we will not cover
> them in depth, but instead focus on how NGINX Kubernetes Gateway supports them.
```

sjberman commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+
+This guide covers:
+
+- Rolling Deployment upgrade;
+- Blue-green Deployments;
+- Canary releases.
```
```suggestion
- Rolling Deployment upgrade
- Blue-green Deployments
- Canary releases
```

sjberman commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+For the upgrade methods covered in the next sections, we make the following assumptions:
+
+- You deploy your application as a [Deployment][deployment].
+- The pods of the Deployment belong to a [Service][service] so that Kubernetes creates an [Endpoint][endpoints] for
+  each pod.
```
```suggestion
- The Pods of the Deployment belong to a [Service][service] so that Kubernetes creates an [Endpoint][endpoints] for
  each Pod.
```

sjberman commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+### Graceful Configuration Reloads
+
+If a relevant Gateway API or Kubernetes built-in resource is changed, NGINX Kubernetes Gateway will update NGINX
+accordingly: it will regenerate NGINX configuration and then send a reload signal to the NGINX master process to apply
```
```suggestion
accordingly by regenerating NGINX configuration and then sending a reload signal to the NGINX master process to apply
```

sjberman commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+
+### Adding or Removing Endpoints
+
+During an upgrade of an application, Kubernetes starts the pods of the new version and brings down the old ones. It also
```
```suggestion
During an upgrade of an application, Kubernetes starts the Pods of the new version and brings down the old ones. It also
```

sjberman commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+
+Let's consider two cases:
+
+- If an Endpoint is added, NGINX Kubernetes Gateway will add an upstream server that corresponds to the Endpoint to
+- NGINX and reload NGINX. After that, NGINX will start proxying traffic to it.
```
```suggestion
- If an Endpoint is added, NGINX Kubernetes Gateway adds an upstream server to NGINX that corresponds to the Endpoint, and then reloads NGINX. After that, NGINX will start proxying traffic to that Endpoint.
```

sjberman commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+
+- If an Endpoint is added, NGINX Kubernetes Gateway will add an upstream server that corresponds to the Endpoint to
+- NGINX and reload NGINX. After that, NGINX will start proxying traffic to it.
+- If an Endpoint is removed, NGINX Kubernetes Gateway will remove the corresponding upstream server from NGINX. After
```
```suggestion
- If an Endpoint is removed, NGINX Kubernetes Gateway removes the corresponding upstream server from NGINX. After
```

sjberman commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+## Rolling Deployment Upgrade
+
+To start a [rolling Deployment upgrade][rolling-upgrade], you update the Deployment to use the new version tag of
+the application. As a result, Kubernetes will terminate the pods with the old
```
```suggestion
the application. As a result, Kubernetes terminates the Pods with the old
```

sjberman commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+
+To start a [rolling Deployment upgrade][rolling-upgrade], you update the Deployment to use the new version tag of
+the application. As a result, Kubernetes will terminate the pods with the old
+version and create new ones, ensuring, by default, that some number of pods always stays available during this process.
```
```suggestion
version and create new ones. This ensures that some number of Pods always stays available during this process.
```

pleshakov replied:
address this in https://github.com/nginxinc/nginx-kubernetes-gateway/pull/944/commits/3622223d78183c328c4424bac13443389efdc961 I retained by default, because this behavior could be changed if the default .spec.strategy.type is changed from to the Recreate

sjberman commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+  Gateway will remove the green upstream servers from NGINX and add the blue ones. With this approach, it is not
+  necessary to update the HTTPRoute.
+- Create a separate Service for the blue version and update the backend reference in the HTTPRoute to reference this
+  Service, which will lead to the same result as with the previous option.
```
```suggestion
  Service, which leads to the same result as with the previous option.
```

sjberman commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+## Canary Releases
+
+To support canary releases, you can implement an approach with two Deployments behind the same Service (see
+[Canary deployment][canary] in the Kubernetes documentation). For example, if you have an old Deployment with four pods
+and a new Deployment with one pod, NGINX will proxy about 80% of client requests to the old pods and 20% to the new pod.
```
```suggestion
[Canary deployment][canary] in the Kubernetes documentation). For example, if you have an old Deployment with four Pods
and a new Deployment with one Pod, NGINX will proxy about 80% of client requests to the old Pods and 20% to the new Pod.
```

sjberman commented on docs/guides/upgrade-apps-without-downtime.md:
````diff
@@ -0,0 +1,182 @@
+    weight: 5
+```
+
+> There is no stickiness for the requests from the same client: NGINX will independently split each request among
````
```suggestion
> There is no stickiness for the requests from the same client. NGINX will independently split each request among
```

sjberman reviewed (commented):
I'd also tag someone from the docs team and PM to do a review of this.

kate-osborn commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+
+To start a [rolling Deployment upgrade][rolling-upgrade], you update the Deployment to use the new version tag of
+the application. As a result, Kubernetes terminates the Pods with the old version and create new ones. By default,
+Kubernetes also ensures that some number of pods always stays available during the upgrade.
```
```suggestion
Kubernetes also ensures that some number of pods always stay available during the upgrade.
```

kate-osborn commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+
+To support canary releases, you can implement an approach with two Deployments behind the same Service (see
+[Canary deployment][canary] in the Kubernetes documentation). For example, if you have an old Deployment with four Pods
+and a new Deployment with one Pod, NGINX will proxy about 80% of client requests to the old Pods and 20% to the new Pod.
```
Is this true given that the lb algorithm we use is `random two least_conn`?

pleshakov replied:
good question!

how about this version:

To support canary releases, you can implement an approach with two Deployments behind the same Service (see
[Canary deployment]() in the Kubernetes documentation). However, this approach lacks precision for defining the traffic split between the old and the canary version. You can greatly influence it by controlling the number of pods (for example, four Pods of the old version and one Pod of the canary). However, note that NGINX Kubernetes Gateway uses [`random two least_conn`](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#random) load balancing method, which doesn't guarantee an 80/20 split based on the number of pods in the given example.

kate-osborn replied:
LGTM!

kate-osborn reviewed (commented):
Great doc! Just a couple small comments

mpstefan commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+- If an Endpoint is removed, NGINX Kubernetes Gateway removes the corresponding upstream server from NGINX. After
+  a reload, NGINX will stop proxying traffic to it. However, it will finish proxying any pending requests to that
+  server,
+  which prevents downtime.
```
Not sure if this was intentional - looks like an unintentional line break?

mpstefan commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,182 @@
+With this method, you deploy a new version of the application (blue version) as a separate Deployment,
+while the old version (green) keeps running and handling client traffic. Next, you switch the traffic from the
+green version to the blue. If the blue works as expected, you terminate the green or switch the traffic back to it
+otherwise.
```
For readability, can we say something like... "If the blue application works as expected, the green application should be terminated. If the blue application does not work as expected, the traffic should be switched back to the green application."

kate-osborn reviewed (approved):
🚀

sjberman commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,174 @@
+
+It ensures that clients attempting to access your application will receive standard response codes (such as `502`) instead of empty responses during upgrades.
+
+Multiple upgrade methods are metnioned, assuming existing familiarity: this guide focuses primarily on how to use NGINX Kubernetes Gateway to accomplish them.
```
```suggestion
Multiple upgrade methods are mentioned, assuming existing familiarity: this guide focuses primarily on how to use NGINX Kubernetes Gateway to accomplish them.
```

pleshakov commented:
@ADubhlaoich 

> I have pushed a large amount of changes to the document which restructure the text and rephrase much of the content. Most of the changes were for the sake of imperative instructions and removing various qualifiers, making the text more concise.

👍 

> Some of these changes, however, do not pass due to the Markdown line length checker. What convention is that specific length based on?

the length limit is 80. We choose it for all markdown files in our repo, so that is is easier read the source file and to review.

I made a few changes based on your changes. Mainly:
- removed "It ensures that clients attempting to access your application will receive standard response codes (such as `502`) instead of empty responses during upgrades." because this is not the goal of the guide -- the goal is to prevent downtime, which are 502 responses.
- renamed "Endpoint switching" to "Upstream Servers Updates" - to make it use NGINX terminology.
- a few smaller changes and fixes

Please let me know if those are ok.

ADubhlaoich reviewed (approved):
Approved with one minor suggestion - "Upstream Servers Updates" => "Upstream Server Updates". Updates is already pluralised and making Server singular is better for readability, with the possibility of multiple implied in context.

If the plan at a later stage is to serve the docs through a website, the markdown line length checking may become an issue since the line length is currently prioritising GitHub readability. 

IDEs and web frameworks have no issue with word wrapping paragraphs of arbitrary line length, so this may become the writing equivalent of technical debt that'll need to be addressed if the default expectation of reader consumption through GitHub changes.

pleshakov commented on docs/guides/upgrade-apps-without-downtime.md:
```diff
@@ -0,0 +1,173 @@
+## NGINX Kubernetes Gateway Functionality
+
+To understand the upgrade methods, you should be aware of the NGINX features that help prevent application downtime:
+graceful configuration reloads and upstream servers updates.
```
```suggestion
graceful configuration reloads and upstream server updates.
```
